drivers: dmic: clear SOFT_RESET bit to ACE as well

fredoh9 opened the pull request:
Clearing sof-reset bit applies to ACE as well. And when releasing a paused stream, unmuting mic is necessary. Without unmuting there will be no input data after resume.

fixes: https://github.com/thesofproject/sof/issues/6723

fredoh9 committed:
drivers: dmic: clear SOFT_RESET bit for ACE as well

Clearing SOFT_RESET bit applies to ACE as well. In case of ACE,
this should be done before setting start bits for CIC and FIR.

Signed-off-by: Fred Oh <fred.oh@linux.intel.com>

ranj063 reviewed (dismissed):
Thanks @fredoh9 !

teburd commented:
nice find @ranj063 @fredoh9 !

dnikodem commented:
> @dnikodem can you run your internal tests with this? I remember this was like this because mtl was not very happy without it...

Sure. I have asked our validation to help with this. I will let you know when the results are ready. Thanks

kv2019i commented on drivers/dai/intel/dmic/dmic.c:
```diff
@@ -607,18 +607,16 @@ static void dai_dmic_start(struct dai_intel_dmic *dmic)
 	for (i = 0; i < CONFIG_DAI_DMIC_HW_CONTROLLERS; i++) {
 		dai_dmic_update_bits(dmic, base[i] + CIC_CONTROL,
-				     CIC_CONTROL_SOFT_RESET_BIT, 0);
+				     CIC_CONTROL_SOFT_RESET_BIT|CIC_CONTROL_MIC_MUTE_BIT, 0);
```
I wonder if it would be better to have two separate commits, one to apply this code on ACE as well, and second commit to add the MUTE bit. 

Minor style note: empty space around bit operations is preferred, i.e. "CIC_CONTROL_SOFT_RESET_BIT | CIC_CONTROL_MIC_MUTE_BIT"

fredoh9 replied:
I'm still working on it. I have some cases are failing today, trying to make sure everything is covered by this patch

kv2019i reviewed (commented):
Good catch @fredoh9 ! Some minor notes inline.

juimonen commented:
> @juimonen @fredoh9 @dnikodem using FW from [thesofproject/sof#7332](https://github.com/thesofproject/sof/pull/7332), I checked the DMIC on MTL and found no regression. Everything seems fine

@wszypelt thanks! then I'm ok with this PR.

plbossart commented on drivers/dai/intel/dmic/dmic.c:
```diff
@@ -607,18 +607,16 @@ static void dai_dmic_start(struct dai_intel_dmic *dmic)
 	 */
 	for (i = 0; i < CONFIG_DAI_DMIC_HW_CONTROLLERS; i++) {
 		dai_dmic_update_bits(dmic, base[i] + CIC_CONTROL,
-				     CIC_CONTROL_SOFT_RESET_BIT, 0);
```
@singalsu can you clarify your comment? it's not clear to me if you are suggesting another solution and based on what?

If the blobs should be changed that's not a zephyr change but a topology plugin change to handled by @juimonen

fredoh9 replied:
@singalsu I tried below diff instead of second commit, but it didn't work. I have only zero input. Can you show me what I missed?

```diff
--- a/drivers/dai/intel/dmic/dmic.c
+++ b/drivers/dai/intel/dmic/dmic.c
@@ -604,12 +604,20 @@ static void dai_dmic_start(struct dai_intel_dmic *dmic)
                switch (dmic->dai_config_params.dai_index) {
                case 0:
                        dai_dmic_update_bits(dmic, base[i] + FIR_CONTROL_A,
+#ifdef CONFIG_SOC_SERIES_INTEL_ACE
+                                            FIR_CONTROL_A_START_BIT | FIR_CONTROL_A_PERIODIC_START_EN_BIT,
+#else
                                             FIR_CONTROL_A_START_BIT,
+#endif
                                             FIR_CONTROL_A_START(fir_a));
                        break;
                case 1:
                        dai_dmic_update_bits(dmic, base[i] + FIR_CONTROL_B,
+#ifdef CONFIG_SOC_SERIES_INTEL_ACE
+                                            FIR_CONTROL_B_START_BIT | FIR_CONTROL_B_PERIODIC_START_EN_BIT,
+#else
                                             FIR_CONTROL_B_START_BIT,
+#endif
                                             FIR_CONTROL_B_START(fir_b));
                        break;
                }
```

fredoh9 committed:
drivers: dmic: clear MIC_MUTE bit for ACE

When releasing a paused stream, clearing MIC_MUTE bit is necessary.
Without unmuting there will be no input data after resume.

Signed-off-by: Fred Oh <fred.oh@linux.intel.com>

kv2019i reviewed (approved):
Excellent, thanks @fredoh9 ! We might want to revisit whether INF is really needed for start, but I think given the amount of issues with the dmic start flows, I think it makes sense to have it in the logs and this anyways is only once per stream start.
